diff -Naur openswan-2.6.32-orig/include/ipsecconf/keywords.h openswan-2.6.32/include/ipsecconf/keywords.h
--- openswan-2.6.32-orig/include/ipsecconf/keywords.h	2013-04-24 13:21:09.628358190 -0400
+++ openswan-2.6.32/include/ipsecconf/keywords.h	2013-04-24 13:22:51.996741775 -0400
@@ -231,6 +231,7 @@
     kt_bitstring,          /* an encryption/authentication key */
     kt_comment,            /* a value which is a cooked comment */
     kt_obsolete,           /* option that is obsoleted, allow keyword but warn and ignore */
+    kt_obsolete_quiet,     /* option that is obsoleted, allow keyword but don't bother warning */
 };
 
 #define NOT_ENUM NULL
diff -Naur openswan-2.6.32-orig/lib/libipsecconf/confread.c openswan-2.6.32/lib/libipsecconf/confread.c
--- openswan-2.6.32-orig/lib/libipsecconf/confread.c	2013-04-24 13:21:09.628358190 -0400
+++ openswan-2.6.32/lib/libipsecconf/confread.c	2013-04-24 13:24:13.157836866 -0400
@@ -323,6 +323,9 @@
 	    case kt_obsolete:
 	        starter_log(LOG_LEVEL_INFO, "Warning: ignored obsolete keyword %s",kw->keyword.keydef->keyname);
 		break;
+	    case kt_obsolete_quiet:
+	        starter_log(LOG_LEVEL_DEBUG, "Warning: ignored obsolete keyword %s",kw->keyword.keydef->keyname);
+		break;
 	    }
 	}
 		
@@ -791,6 +794,9 @@
 	case kt_comment:
 	    break;
 	case kt_obsolete:
+	    starter_log(LOG_LEVEL_INFO,"Warning: obsolete keyword %s ignored\n",kw->keyword.keydef->keyname);
+	    break;
+	case kt_obsolete_quiet:
 	    starter_log(LOG_LEVEL_DEBUG,"Warning: obsolete keyword %s ignored\n",kw->keyword.keydef->keyname);
 	    break;
 	}
diff -Naur openswan-2.6.32-orig/lib/libipsecconf/confwrite.c openswan-2.6.32/lib/libipsecconf/confwrite.c
--- openswan-2.6.32-orig/lib/libipsecconf/confwrite.c	2013-04-24 13:21:09.607357906 -0400
+++ openswan-2.6.32/lib/libipsecconf/confwrite.c	2013-04-24 13:22:51.998741802 -0400
@@ -153,6 +153,7 @@
 	    continue;
 
 	case kt_obsolete:
+	case kt_obsolete_quiet:
 	    continue;
 	}
 
@@ -224,6 +225,7 @@
 	    continue;
 
 	case kt_obsolete:
+	case kt_obsolete_quiet:
 	    continue;
 	}
 
diff -Naur openswan-2.6.32-orig/lib/libipsecconf/keywords.c openswan-2.6.32/lib/libipsecconf/keywords.c
--- openswan-2.6.32-orig/lib/libipsecconf/keywords.c	2013-04-24 13:21:09.628358190 -0400
+++ openswan-2.6.32/lib/libipsecconf/keywords.c	2013-04-24 13:24:51.796357640 -0400
@@ -359,7 +359,7 @@
     {"manualstart",    kv_config, kt_string,    KSF_MANUALSTART,NOT_ENUM},
     {"pluto",          kv_config, kt_filename,  KSF_PLUTO, NOT_ENUM},
     {"plutowait",      kv_config, kt_bool,      KBF_PLUTOWAIT,NOT_ENUM},
-    {"oe",             kv_config, kt_bool,      KBF_OPPOENCRYPT,NOT_ENUM},
+    {"oe",             kv_config, kt_obsolete_quiet,      KBF_WARNIGNORE,NOT_ENUM},
     {"prepluto",       kv_config, kt_filename,  KSF_PREPLUTO,NOT_ENUM},
     {"postpluto",      kv_config, kt_filename,  KSF_POSTPLUTO,NOT_ENUM},
     {"fragicmp",       kv_config, kt_bool,      KBF_FRAGICMP,NOT_ENUM},
diff -Naur openswan-2.6.32-orig/lib/libipsecconf/parser.y openswan-2.6.32/lib/libipsecconf/parser.y
--- openswan-2.6.32-orig/lib/libipsecconf/parser.y	2010-12-17 20:23:54.000000000 -0500
+++ openswan-2.6.32/lib/libipsecconf/parser.y	2013-04-24 13:22:51.998741802 -0400
@@ -195,6 +195,7 @@
                         break;
 
            	    case kt_obsolete:
+           	    case kt_obsolete_quiet:
                         break;
 		    }	
 		    new->next = NULL;
@@ -265,6 +266,7 @@
            	    case kt_comment:
                         break;
            	    case kt_obsolete:
+           	    case kt_obsolete_quiet:
                         break;
 		    }	
 		    new->next = NULL;
