diff -Naur openswan-2.6.32-orig/Makefile.inc openswan-2.6.32/Makefile.inc
--- openswan-2.6.32-orig/Makefile.inc	2013-10-17 13:06:53.394234083 -0400
+++ openswan-2.6.32/Makefile.inc	2013-10-17 13:07:28.734927924 -0400
@@ -353,6 +353,7 @@
 
 # Support for integrity check for binaries (requires USE_LIBNSS and fipscheck-devel)
 USE_FIPSCHECK?=false
+FIPSPRODUCTCHECK?=/etc/system-fips
 
 # Support for NSS crypto library (does not requires HAVE_THREADS)
 # USE_LIBNSS uses pthreads by default.
diff -Naur openswan-2.6.32-orig/Makefile.top openswan-2.6.32/Makefile.top
--- openswan-2.6.32-orig/Makefile.top	2013-10-17 13:06:53.270231648 -0400
+++ openswan-2.6.32/Makefile.top	2013-10-17 13:07:28.734927924 -0400
@@ -110,5 +110,5 @@
 export USE_TAPROOM USE_OBJDIR
 export HAVE_STATSD USE_DYNAMICDNS
 export USE_IPSEC_CONNECTION_LIMIT IPSEC_CONNECTION_LIMIT
-export USE_LIBNSS USE_FIPSCHECK USE_MODP_RFC5114 USE_NM USE_LABELED_IPSEC
+export USE_LIBNSS USE_FIPSCHECK FIPSPRODUCTCHECK USE_MODP_RFC5114 USE_NM USE_LABELED_IPSEC
 export USE_MAST USE_SAREF_KERNEL
diff -Naur openswan-2.6.32-orig/programs/pluto/Makefile openswan-2.6.32/programs/pluto/Makefile
--- openswan-2.6.32-orig/programs/pluto/Makefile	2013-10-17 13:06:53.395234103 -0400
+++ openswan-2.6.32/programs/pluto/Makefile	2013-10-17 13:07:28.734927924 -0400
@@ -43,7 +43,8 @@
 CPPFLAGS = $(HDRDIRS) $(DEFINES) \
 	-DSHARED_SECRETS_FILE=\"${FINALCONFDIR}/ipsec.secrets\" \
 	-DPOLICYGROUPSDIR=\"${FINALCONFDDIR}/policies\" \
-	-DPERPEERLOGDIR=\"${FINALLOGDIR}/pluto/peer\"
+	-DPERPEERLOGDIR=\"${FINALLOGDIR}/pluto/peer\" \
+	-DFIPSPRODUCTCHECK=\"${FIPSPRODUCTCHECK}\" \
 
 ifeq ($(HAVE_BROKEN_POPEN),true)
 CFLAGS+=-DHAVE_BROKEN_POPEN
diff -Naur openswan-2.6.32-orig/programs/pluto/plutomain.c openswan-2.6.32/programs/pluto/plutomain.c
--- openswan-2.6.32-orig/programs/pluto/plutomain.c	2013-10-17 13:06:53.386233926 -0400
+++ openswan-2.6.32/programs/pluto/plutomain.c	2013-10-17 14:26:25.489053572 -0400
@@ -286,6 +286,26 @@
     }
 }
 
+#ifdef FIPS_CHECK
+static bool
+osw_fipsproduct(void)
+{
+	if (access(FIPSPRODUCTCHECK, F_OK) != 0) {
+		if (errno == ENOENT || errno == ENOTDIR) {
+			openswan_log("FIPS: not a FIPS product");
+			return FALSE; 
+		} else {
+			loglog(RC_LOG_SERIOUS, "FIPS ABORT: FIPS product check"
+				" failed to determine status: %d: %s", errno,
+				strerror(errno));
+			exit_pluto(1);
+			return FALSE; /* make compiler happy - never reached */
+		}
+	}
+	return TRUE;
+}
+#endif
+
 /** by default pluto sends certificate requests to its peers */
 bool no_cr_send = FALSE;
 
@@ -860,6 +880,10 @@
 	    PK11_SetPasswordFunc(getNSSPassword);
 
 #ifdef FIPS_CHECK
+	{
+
+	int fips_mode = Pluto_IsFIPS();
+	int fips_product = osw_fipsproduct();
 	const char *package_files[]= { IPSECLIBDIR"/setup",
 				        IPSECLIBDIR"/addconn",
 				        IPSECLIBDIR"/auto",
@@ -899,13 +923,35 @@
 					IPSECSBINDIR"/ipsec",
 					NULL
 					};
+	int fips_files_check_ok = FIPSCHECK_verify_files(package_files);
 
-       if (Pluto_IsFIPS() && !FIPSCHECK_verify_files(package_files)) {
-             loglog(RC_LOG_SERIOUS, "FIPS integrity verification test failed");
-             exit_pluto(10);
-        }
-#endif
+	if (fips_product)
+		openswan_log("FIPS Product detected (%s)", FIPSPRODUCTCHECK);
 
+	if (fips_mode)
+		openswan_log("FIPS Kernel Mode detected");
+
+	if (!fips_files_check_ok) {
+			loglog(RC_LOG_SERIOUS, "FIPS HMAC integrity verification FAILURE");
+		/*
+		 * We ignore fips=1 kernel mode if we have no fips product installed
+		 */
+		if (fips_product && fips_mode) {
+			loglog(RC_LOG_SERIOUS, "ABORT: FIPS product and kernel in FIPS mode");
+			exit_pluto(10);
+		} else if (fips_product) {
+			openswan_log("FIPS: FIPS product but kernel mode disabled - continuing");
+		} else {
+			openswan_log("FIPS: not a FIPS product, kernel mode ignored - continuing");
+		}
+	} else {
+		openswan_log("FIPS HMAC integrity verification test passed");
+	}
+
+	}
+#else
+	openswan_log("FIPS HMAC integrity support [disabled]");
+#endif
       }
 #endif
 
diff -Naur openswan-2.6.32-orig/lib/libopenswan/oswconf.c openswan-2.6.32/lib/libopenswan/oswconf.c
--- openswan-2.6.32-orig/lib/libopenswan/oswconf.c	2010-12-17 20:23:54.000000000 -0500
+++ openswan-2.6.32/lib/libopenswan/oswconf.c	2013-10-17 14:35:57.833319440 -0400
@@ -179,6 +179,7 @@
      
      if(fd!=NULL) {
 	    n = fread ((void *)fips_flag, 1, 1, fd);
+	    fclose(fd); 
 		if(n==1) {
 		    if(fips_flag[0]=='1') {
 	            return TRUE;	         	
