diff -Naur openswan-2.6.32-orig/programs/_startnetkey/_startnetkey.in openswan-2.6.32/programs/_startnetkey/_startnetkey.in
--- openswan-2.6.32-orig/programs/_startnetkey/_startnetkey.in	2010-12-17 20:23:54.000000000 -0500
+++ openswan-2.6.32/programs/_startnetkey/_startnetkey.in	2013-04-27 19:41:55.051738539 -0400
@@ -197,7 +197,7 @@
 
 		# load the most common ciphers/algo's
 		# aes-x86_64 has higher priority in via crypto api
-		for crypto in aes-x86_64 geode-aes aes aes_generic des sha512 sha256 md5 cbc xcbc ecb twofish blowfish serpent ccm
+		for crypto in aesni_intel aes-x86_64 geode-aes aes aes_generic des sha512 sha256 md5 cbc xcbc ecb twofish blowfish serpent ccm
 		   do
 			#echo -n "$crypto "
 			@MODPROBE@ $crypto 2> /dev/null
