diff -urNp openswan-2.6.32-cvs-patched/programs/pluto/kernel_netlink.c openswan-2.6.32-current/programs/pluto/kernel_netlink.c
--- openswan-2.6.32-cvs-patched/programs/pluto/kernel_netlink.c	2011-04-04 15:59:33.769214787 -0400
+++ openswan-2.6.32-current/programs/pluto/kernel_netlink.c	2011-05-12 17:39:35.126820075 -0400
@@ -894,7 +894,7 @@ netlink_add_sa(struct kernel_sa *sa, boo
 	req.n.nlmsg_len += attr->rta_len;
 	attr = (struct rtattr *)((char *)attr + attr->rta_len);
     }
-    else
+    else if (sa->esatype == ET_ESP)
     {
 	struct xfrm_algo algo;
 	const char *name;
