diff -urNp openswan-2.6.32-orig/programs/ipsec/ipsec.in openswan-2.6.32-cvs-patched/programs/ipsec/ipsec.in
--- openswan-2.6.32-orig/programs/ipsec/ipsec.in	2010-12-20 12:44:19.150080076 -0500
+++ openswan-2.6.32-cvs-patched/programs/ipsec/ipsec.in	2010-12-20 12:55:34.269071757 -0500
@@ -80,9 +80,9 @@ case "$1" in
 --help)
 	echo "Usage: ipsec command argument ..."
 	echo "where command is one of:"
-	for f in `ls $IPSEC_LIBDIR $IPSEC_EXECDIR | egrep -v -i "$DONTMENTION"`
+	for f in `ls $IPSEC_LIBDIR | egrep -v -i "$DONTMENTION"`
 	do
-		if test -x $IPSEC_LIBDIR/$f || test -x $IPSEC_EXECDIR/$f
+		if test -x $IPSEC_LIBDIR/$f
 		then
 			echo "	$f"
 		fi
