diff -Naur openswan-2.6.32-orig/include/ipsecconf/keywords.h openswan-2.6.32/include/ipsecconf/keywords.h
--- openswan-2.6.32-orig/include/ipsecconf/keywords.h	2013-04-07 21:36:44.112000117 -0400
+++ openswan-2.6.32/include/ipsecconf/keywords.h	2013-04-07 21:37:13.870999539 -0400
@@ -112,6 +112,7 @@
     KBF_WARNIGNORE, /* to ignore obsoleted keywords */
     KBF_SECCTX, /*security context attribute value for labeled ipsec*/
     KBF_IKE_FRAG,
+    KBF_NAT_KEEPALIVE, /* per conn enabling/disabling of sending keep-alives - different from global force_keepalives */
     KBF_MAX         
 };
 
diff -Naur openswan-2.6.32-orig/include/whack.h openswan-2.6.32/include/whack.h
--- openswan-2.6.32-orig/include/whack.h	2013-04-07 21:36:43.874000102 -0400
+++ openswan-2.6.32/include/whack.h	2013-04-07 21:37:13.871999576 -0400
@@ -132,6 +132,9 @@
     /* Force the use of NAT-T on a connection */
     bool forceencaps;
 
+    /* Option to allow per-conn setting of sending of NAT-T keepalives - default is enabled  */
+    bool nat_keepalive;
+
     enum keyword_sha2_truncbug sha2_truncbug;
 
     /*Checking if this connection is configured by Network Manager*/
diff -Naur openswan-2.6.32-orig/lib/libipsecconf/confread.c openswan-2.6.32/lib/libipsecconf/confread.c
--- openswan-2.6.32-orig/lib/libipsecconf/confread.c	2013-04-07 21:36:44.114000114 -0400
+++ openswan-2.6.32/lib/libipsecconf/confread.c	2013-04-07 21:37:13.871999576 -0400
@@ -71,6 +71,7 @@
 	cfg->setup.options[KBF_DISABLEPORTFLOATING]= FALSE;
 	cfg->setup.options[KBF_FORCE_KEEPALIVE]= FALSE;
 	cfg->setup.options[KBF_KEEPALIVE]= 0;
+	cfg->conn_default.options[KBF_NAT_KEEPALIVE] = TRUE; /* per conn */
 #endif
 	cfg->conn_default.options[KBF_TYPE] = KS_TUNNEL;
 
diff -Naur openswan-2.6.32-orig/lib/libipsecconf/keywords.c openswan-2.6.32/lib/libipsecconf/keywords.c
--- openswan-2.6.32-orig/lib/libipsecconf/keywords.c	2013-04-07 21:36:44.115000114 -0400
+++ openswan-2.6.32/lib/libipsecconf/keywords.c	2013-04-07 21:37:13.872999611 -0400
@@ -422,6 +422,7 @@
     {"sareftrack",     kv_conn|kv_auto|kv_processed,kt_enum,KBF_SAREFTRACK,&kw_sareftrack_list},
     {"ike_frag",       kv_conn|kv_auto|kv_processed,kt_enum,KBF_IKE_FRAG,&kw_ynf_list},
     {"pfs",            kv_conn|kv_auto, kt_bool,   KBF_PFS,          NOT_ENUM},
+    {"nat_keepalive",  kv_conn|kv_auto, kt_bool,   KBF_NAT_KEEPALIVE,          NOT_ENUM},
     {"sha2_truncbug",  kv_conn|kv_auto, kt_enum,   KBF_SHA2_TRUNCBUG,          &kw_sha2_truncbug},
     {"keylife",        kv_conn|kv_auto|kv_alias, kt_time,   KBF_SALIFETIME,NOT_ENUM},
     {"lifetime",       kv_conn|kv_auto|kv_alias, kt_time,   KBF_SALIFETIME,NOT_ENUM},
diff -Naur openswan-2.6.32-orig/lib/libipsecconf/starterwhack.c openswan-2.6.32/lib/libipsecconf/starterwhack.c
--- openswan-2.6.32-orig/lib/libipsecconf/starterwhack.c	2013-04-07 21:36:43.876000110 -0400
+++ openswan-2.6.32/lib/libipsecconf/starterwhack.c	2013-04-07 21:37:13.872999611 -0400
@@ -514,6 +514,9 @@
 	if(conn->options_set[KBF_FORCEENCAP]) {
 		msg.forceencaps=conn->options[KBF_FORCEENCAP];
 	}
+	if(conn->options_set[KBF_NAT_KEEPALIVE]) {
+		msg.nat_keepalive=conn->options[KBF_NAT_KEEPALIVE];
+	} else msg.nat_keepalive = TRUE;
 #endif
 
 	/*Cisco interop : remote peer type*/
diff -Naur openswan-2.6.32-orig/programs/_confread/d.ipsec.conf/nat_keepalive.xml openswan-2.6.32/programs/_confread/d.ipsec.conf/nat_keepalive.xml
--- openswan-2.6.32-orig/programs/_confread/d.ipsec.conf/nat_keepalive.xml	1969-12-31 19:00:00.000000000 -0500
+++ openswan-2.6.32/programs/_confread/d.ipsec.conf/nat_keepalive.xml	2013-04-07 21:37:13.872999611 -0400
@@ -0,0 +1,12 @@
+  <varlistentry>
+  <term><emphasis remap='B'>nat_keepalive</emphasis></term>
+  <listitem>
+<para>whether to send any NAT-T keep-alives. These 1 byte packets 
+are send to prevent the NAT router from closing its port when there is
+not enough traffic on the IPsec connection. Acceptable values are:
+<emphasis remap='B'>yes</emphasis> (the deault) and
+<emphasis remap='B'>no</emphasis>. This per-conn option takes precedence
+over the global (config setup) <emphasis remap='I'>force_keepalive=</emphasis> option.
+</para>
+  </listitem>
+  </varlistentry>
diff -Naur openswan-2.6.32-orig/programs/_confread/d.ipsec.conf/order.txt openswan-2.6.32/programs/_confread/d.ipsec.conf/order.txt
--- openswan-2.6.32-orig/programs/_confread/d.ipsec.conf/order.txt	2013-04-07 21:36:43.877000116 -0400
+++ openswan-2.6.32/programs/_confread/d.ipsec.conf/order.txt	2013-04-07 21:37:13.872999611 -0400
@@ -40,6 +40,7 @@
 d.ipsec.conf/remote_peer_type.xml
 d.ipsec.conf/nm_configured.xml
 d.ipsec.conf/forceencaps.xml
+d.ipsec.conf/nat_keepalive.xml
 d.ipsec.conf/overlapip.xml
 d.ipsec.conf/dpddelay.xml
 d.ipsec.conf/dpdtimeout.xml
diff -Naur openswan-2.6.32-orig/programs/pluto/connections.c openswan-2.6.32/programs/pluto/connections.c
--- openswan-2.6.32-orig/programs/pluto/connections.c	2013-04-07 21:36:43.879000117 -0400
+++ openswan-2.6.32/programs/pluto/connections.c	2013-04-07 21:38:30.332000138 -0400
@@ -1360,6 +1360,7 @@
 	c->metric = wm->metric;
 
 	c->forceencaps = wm->forceencaps;
+	c->nat_keepalive = wm->nat_keepalive;
 
 	c->addr_family = wm->addr_family;
 	c->tunnel_addr_family = wm->tunnel_addr_family;
@@ -3344,7 +3345,7 @@
     
     whack_log(RC_COMMENT
 	      , "\"%s\"%s:   ike_life: %lus; ipsec_life: %lus;"
-	      " rekey_margin: %lus; rekey_fuzz: %lu%%; keyingtries: %lu%s "
+	      " rekey_margin: %lus; rekey_fuzz: %lu%%; keyingtries: %lu%s%s "
 	      , c->name
 	      , instance
 	      , (unsigned long) c->sa_ike_life_seconds
@@ -3353,6 +3354,7 @@
 	      , (unsigned long) c->sa_rekey_fuzz
 	      , (unsigned long) c->sa_keying_tries
 	      , (c->sha2_truncbug) ? "; sha2_truncbug: yes" : ""
+	      , (c->nat_keepalive) ? "; nat_keepalive: yes" : ""
 	     );
 
     if (c->policy_next)
@@ -3377,12 +3379,13 @@
     /* slightly complicated stuff to avoid extra crap */
     if(c->dpd_timeout > 0 || DBGP(DBG_DPD)) {
 	whack_log(RC_COMMENT
-		  , "\"%s\"%s:   dpd: %s; delay:%lu; timeout:%lu; "
+		  , "\"%s\"%s:   dpd: %s; delay:%lu; timeout:%lu;"
 		  , c->name
 		  , instance
 		  , enum_name(&dpd_action_names, c->dpd_action)
 		  , (unsigned long)c->dpd_delay
-		  , (unsigned long)c->dpd_timeout);
+		  , (unsigned long)c->dpd_timeout
+		  );
     }
 
     if(c->extra_debugging) {
diff -Naur openswan-2.6.32-orig/programs/pluto/connections.h openswan-2.6.32/programs/pluto/connections.h
--- openswan-2.6.32-orig/programs/pluto/connections.h	2013-04-07 21:36:43.880000116 -0400
+++ openswan-2.6.32/programs/pluto/connections.h	2013-04-07 21:37:13.874999676 -0400
@@ -202,6 +202,8 @@
     time_t          dpd_timeout;            /* time after which we are dead */
     enum dpd_action dpd_action;             /* what to do when we die */
 
+    bool	    nat_keepalive;      /* Suppress sending NAT-T Keep-Alives */
+
     /*Cisco interop: remote peer type*/
     enum keyword_remotepeertype remotepeertype;
 
diff -Naur openswan-2.6.32-orig/programs/pluto/nat_traversal.c openswan-2.6.32/programs/pluto/nat_traversal.c
--- openswan-2.6.32-orig/programs/pluto/nat_traversal.c	2013-04-07 21:36:44.124000115 -0400
+++ openswan-2.6.32/programs/pluto/nat_traversal.c	2013-04-07 21:37:13.875999706 -0400
@@ -353,6 +353,11 @@
 	    st->hidden_variables.st_nat_traversal |= LELEM(NAT_TRAVERSAL_NAT_BHND_ME);
 	    st->hidden_variables.st_natd = md->sender;
 	}
+
+	if(st->st_connection->nat_keepalive) {
+	    DBG(DBG_NATT,
+		DBG_log("NAT_TRAVERSAL nat_keepalive enabled"));
+	}
 }
 
 bool nat_traversal_add_natd(u_int8_t np, pb_stream *outs,
@@ -784,6 +789,17 @@
 	unsigned int *_kap_st = (unsigned int *)data;
 	const struct connection *c = st->st_connection;
 	if (!c) return;
+
+	if(c->nat_keepalive == FALSE) {
+		DBG(DBG_NATT,DBG_log("Suppressing sending of NAT-T KEEP-ALIVE by per-conn configuration (nat_keepalive=no)"));
+		return;
+	}
+	DBG(DBG_NATT,DBG_log("Sending of NAT-T KEEP-ALIVE enabled by per-conn configuration (nat_keepalive=yes)"));
+
+	if(_force_ka) {
+		DBG(DBG_NATT,DBG_log("Sending of NAT-T KEEP-ALIVE forced by global configuration (force_keepalive=yes)"));
+	}
+
 	if ( ((st->st_state == STATE_MAIN_R3)
 	      || (st->st_state == STATE_MAIN_I4)
 	      || (st->st_state == STATE_AGGR_R2)
diff -Naur openswan-2.6.32-orig/programs/pluto/whack.c openswan-2.6.32/programs/pluto/whack.c
--- openswan-2.6.32-orig/programs/pluto/whack.c	2013-04-07 21:36:44.129000115 -0400
+++ openswan-2.6.32/programs/pluto/whack.c	2013-04-07 21:37:13.876999734 -0400
@@ -127,7 +127,7 @@
 	    " \\\n   "
 	    " [--dontrekey]"
 	    " [--aggrmode]"
-	    " [--forceencaps]"
+	    " [--forceencaps] [--no-nat_keepalive]"
             " \\\n   "
             " [--dpddelay <seconds> --dpdtimeout <seconds>]"
             " [--dpdaction (clear|hold|restart|restart_by_peer)]"
@@ -483,6 +483,7 @@
     CD_DPDTIMEOUT,
     CD_DPDACTION,
     CD_FORCEENCAPS,
+    CD_NO_NAT_KEEPALIVE,
     CD_IKE,
     CD_PFSGROUP,
     CD_REMOTEPEERTYPE,
@@ -668,6 +669,7 @@
 	, CD_FAIL0 + (POLICY_FAIL_REJECT >> POLICY_FAIL_SHIFT << AUX_SHIFT) + OO },
     { "dontrekey", no_argument, NULL, CD_DONT_REKEY + OO },
     { "forceencaps", no_argument, NULL, CD_FORCEENCAPS + OO },
+    { "no-nat_keepalive", no_argument, NULL, CD_NO_NAT_KEEPALIVE + OO },
     { "dpddelay", required_argument, NULL, CD_DPDDELAY + OO + NUMERIC_ARG },
     { "dpdtimeout", required_argument, NULL, CD_DPDTIMEOUT + OO + NUMERIC_ARG },
     { "dpdaction", required_argument, NULL, CD_DPDACTION + OO },
@@ -1498,6 +1500,10 @@
             msg.forceencaps = TRUE;
             continue;
 
+	case CD_NO_NAT_KEEPALIVE: /* --no-nat_keepalive */
+            msg.nat_keepalive = FALSE;
+            continue;
+
         case CD_DPDDELAY:
             msg.dpd_delay = opt_whole;
             continue;
