diff -Naur openswan-2.6.32-orig/linux/net/ipsec/addrtot.c openswan-2.6.32/linux/net/ipsec/addrtot.c
--- openswan-2.6.32-orig/linux/net/ipsec/addrtot.c	2010-12-17 20:23:54.000000000 -0500
+++ openswan-2.6.32/linux/net/ipsec/addrtot.c	2013-05-19 14:23:39.551519104 -0400
@@ -79,7 +79,7 @@
 	if (n == 0) {
 	bad:
 	  dst[0]='\0';
-	  strncat(dst, "<invalid>", dstlen);
+	  strncat(dst, "<invalid>", dstlen - 1);
 	  return sizeof("<invalid>");
 	}
 
diff -Naur openswan-2.6.32-orig/programs/pluto/ike_alg.c openswan-2.6.32/programs/pluto/ike_alg.c
--- openswan-2.6.32-orig/programs/pluto/ike_alg.c	2013-05-19 14:04:40.425470525 -0400
+++ openswan-2.6.32/programs/pluto/ike_alg.c	2013-05-19 14:18:19.263007666 -0400
@@ -84,7 +84,7 @@
 		/* failure: encrypt algo must be present */
 		snprintf(ugh_buf, ugh_buf_len, "encrypt algo not found");
 		ret = FALSE;
-	} if ((key_len) && ((key_len < enc_desc->keyminlen) ||
+	} else if ((key_len) && ((key_len < enc_desc->keyminlen) ||
 			 (key_len > enc_desc->keymaxlen))) {
 		/* failure: if key_len specified, it must be in range */
 		snprintf(ugh_buf, ugh_buf_len,
