diff -urNp openswan-2.6.32-orig/programs/pluto/plutomain.c openswan-2.6.32-cvs-patched/programs/pluto/plutomain.c
--- openswan-2.6.32-orig/programs/pluto/plutomain.c	2010-12-17 20:23:54.000000000 -0500
+++ openswan-2.6.32-cvs-patched/programs/pluto/plutomain.c	2011-01-12 12:52:44.169048810 -0500
@@ -688,11 +688,8 @@ main(int argc, char **argv)
 
 #ifdef HAVE_NO_FORK
 	fork_desired = FALSE;
+	nhelpers = 0;
 #endif
-    /* if --nofork was given, also disable nhelpers */
-    if (fork_desired == FALSE) {
-           nhelpers = 0;
-    }
 
     /* if a core dir was set, chdir there */
     if(coredir) 
