diff -urNp openswan-2.6.32-cvs-patched/programs/barf/barf.in openswan-2.6.32-current/programs/barf/barf.in
--- openswan-2.6.32-cvs-patched/programs/barf/barf.in	2011-09-01 16:58:25.414675275 -0400
+++ openswan-2.6.32-current/programs/barf/barf.in	2011-10-03 13:38:09.060888069 -0400
@@ -297,13 +297,22 @@ else
 	fi
 fi
 _________________________ iptables
-if test -r /sbin/iptables
+if test -r /sbin/iptables-save
 then
-	iptables -L -v -n
-_________________________ iptables-nat
-	iptables -t nat -L -v -n
-_________________________ iptables-mangle
-	iptables -t mangle -L -v -n
+	iptables-save
+	_________________________ iptables-nat
+	iptables-save -t nat
+	_________________________ iptables-mangle
+	iptables-save -t mangle
+else
+	if test -r /sbin/iptables
+	then
+		iptables -L -v -n
+		_________________________ iptables-nat
+		iptables -t nat -L -v -n
+		_________________________ iptables-mangle
+		iptables -t mangle -L -v -n
+	fi
 fi
 _________________________ /proc/modules
 if test -f /proc/modules
diff -urNp openswan-2.6.32-cvs-patched/programs/verify/verify.in openswan-2.6.32-current/programs/verify/verify.in
--- openswan-2.6.32-cvs-patched/programs/verify/verify.in	2011-09-01 16:58:25.414675275 -0400
+++ openswan-2.6.32-current/programs/verify/verify.in	2011-09-09 18:20:41.459055790 -0400
@@ -401,7 +401,9 @@ sub tunnelchecks {
 	{
 	    errchk "@out";
 	    printfun "Checking NAT and MASQUERADEing";
-	    if( -e "/proc/net/ip_conntrack" )
+	    # The following check is not relevent to NETKEY
+	    # Newer kernel have nf_conntrack not ip_conntrack
+	    if(( -e "/proc/net/nf_conntrack" || -e "/proc/net/ip_conntrack") && -e "/proc/net/ipsec_eroute" )
 	    {
 		run "iptables -t nat -L -n";
 		if(grep /(NAT|MASQ)/, @out)
@@ -421,7 +423,7 @@ sub tunnelchecks {
 	    }
 	    else
 	    { 
-		errchk "","N/A";
+		errchk "OK";
 	    }
 	}
     }
