diff -Naur openswan-2.6.32-orig/programs/ikeping/ikeping.c openswan-2.6.32/programs/ikeping/ikeping.c
--- openswan-2.6.32-orig/programs/ikeping/ikeping.c	2013-02-13 15:42:03.972370424 -0500
+++ openswan-2.6.32/programs/ikeping/ikeping.c	2013-02-13 15:42:58.939372514 -0500
@@ -12,7 +12,6 @@
  * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
  * for more details.
  *
- * RCSID $Id: ikeping.c,v 1.13 2005/07/08 02:56:38 paul Exp $
  */
 
 #include <stdio.h>
@@ -195,7 +194,7 @@
 	unsigned int sendlen;
 	const char *xchg_name;
 	int xchg;
-	u_int32_t tmp_ic, tmp_rc;
+	u_int32_t tmp_ic[2], tmp_rc[2];
 
 	rport = 500;
 	xchg  = 0;
@@ -255,13 +254,14 @@
 	printf("received %d(%s) packet from %s/%d of len: %d\n",
 	       ih.isa_xchg, xchg_name, buf, ntohs(rport), n);
 
-	memcpy(&tmp_ic, ih.isa_icookie, sizeof(u_int32_t));
-	memcpy(&tmp_rc, ih.isa_rcookie, sizeof(u_int32_t));
+	/* questionable: printing each cookie as if it were two uint32 values in host order */
+	memcpy(&tmp_ic, ih.isa_icookie, 2 * sizeof(u_int32_t));
+	memcpy(&tmp_rc, ih.isa_rcookie, 2 * sizeof(u_int32_t));
 	printf("\trcookie=%08x_%08x icookie=%08x_%08x msgid=%08x\n",
-	       tmp_ic,
-	       *(u_int32_t *)(ih.isa_icookie+4), 
-	       tmp_rc,
-	       *(u_int32_t *)(ih.isa_rcookie+4),
+	       tmp_ic[0],
+	       tmp_ic[1],
+	       tmp_rc[0],
+	       tmp_rc[1],
 	       ih.isa_msgid);
 	printf("\tnp=%03d  version=%d.%d    xchg=%s(%d)\n",
 	       ih.isa_np,
