diff -Naur openswan-2.6.32-orig/lib/libopenswan/alg_info.c openswan-2.6.32/lib/libopenswan/alg_info.c
--- openswan-2.6.32-orig/lib/libopenswan/alg_info.c	2013-08-13 14:00:28.180810108 -0400
+++ openswan-2.6.32/lib/libopenswan/alg_info.c	2013-08-13 14:02:48.781391507 -0400
@@ -635,12 +635,6 @@
 		goto out;
 	    }
 
-#ifdef HAVE_LIBNSS
-            if ( Pluto_IsFIPS() && ((aalg_id == OAKLEY_SHA2_256 ) ||(aalg_id == OAKLEY_SHA2_384 ) || (aalg_id == OAKLEY_SHA2_512 ))  ) {
-                p_ctx->err="SHA2 Not supported in FIPS mode with NSS";
-                goto out;
-            }
-#endif
 	    DBG(DBG_CRYPT, DBG_log("parser_alg_info_add() "
 				   "aalg_getbyname(\"%s\")=%d",
 				   p_ctx->aalg_buf,
