diff -Naur openswan-2.6.32-orig/programs/pluto/ikev2_crypto.c openswan-2.6.32/programs/pluto/ikev2_crypto.c
--- openswan-2.6.32-orig/programs/pluto/ikev2_crypto.c	2010-12-17 20:23:54.000000000 -0500
+++ openswan-2.6.32/programs/pluto/ikev2_crypto.c	2013-08-13 21:47:36.181114856 -0400
@@ -65,9 +65,7 @@
 
 	passert(ipi->attrs.transattrs.ei != NULL);
 	memset(&childsacalc, 0, sizeof(childsacalc));
-	childsacalc.prf_hasher = (struct hash_desc *)
-		ike_alg_ikev2_find(IKE_ALG_HASH
-				   , IKEv2_PRF_HMAC_SHA1, 0);
+	childsacalc.prf_hasher = st->st_oakley.prf_hasher;
 	
 	setchunk(childsacalc.ni, st->st_ni.ptr, st->st_ni.len);
 	setchunk(childsacalc.nr, st->st_nr.ptr, st->st_nr.len);
