diff -urNp openswan-2.6.32-patched/programs/pluto/kernel_netlink.c openswan-2.6.32-current/programs/pluto/kernel_netlink.c
--- openswan-2.6.32-patched/programs/pluto/kernel_netlink.c	2012-06-22 12:56:02.983179809 -0400
+++ openswan-2.6.32-current/programs/pluto/kernel_netlink.c	2012-07-05 13:00:55.912477412 -0400
@@ -815,6 +815,7 @@ netlink_add_sa(struct kernel_sa *sa, boo
     {
 	req.p.mode = XFRM_MODE_TRANSPORT;
 
+	if(!sameaddr(sa->src, sa->dst)) {
 	req.p.sel.sport = portof(&sa->src_client->addr);
 	req.p.sel.dport = portof(&sa->dst_client->addr);
 
@@ -851,6 +852,7 @@ netlink_add_sa(struct kernel_sa *sa, boo
 	req.p.sel.prefixlen_d = sa->dst_client->maskbits;
 	req.p.sel.proto = sa->transport_proto;
 	req.p.sel.family = sa->src_client->addr.u.v4.sin_family;
+	}
 
     }
 
