diff -Naur openswan-2.6.32-orig/lib/libopenswan/x509dn.c openswan-2.6.32/lib/libopenswan/x509dn.c
--- openswan-2.6.32-orig/lib/libopenswan/x509dn.c	2013-05-17 22:30:46.085511083 -0400
+++ openswan-2.6.32/lib/libopenswan/x509dn.c	2013-05-17 22:36:35.722001134 -0400
@@ -739,6 +739,39 @@
 	return dntoa(dst, dstlen, dn);
 }
 
+/*
+ * function for use in atodn() that removes a single comma
+ * when encountering ",," in the leftid/rightid string. It
+ * will do this only once for the first ",," found.
+ */
+
+void
+remove_comma(char *str)
+{
+    char comma = ',';
+    char *src, *dst;
+    int once = 0;
+
+    src = dst = str;
+    while (*src != '\0' && once == 0)
+    {
+        *dst = *src;
+        src++;
+        if (*dst++ == comma && *src == comma)
+        {
+            src++;
+            once = 1;
+        }
+    }
+    while (*src != '\0')
+    {
+        *dst = *src;
+        src++;
+        dst++;
+    }
+    *dst = '\0';
+}
+
 /*  Converts an LDAP-style human-readable ASCII-encoded
  *  ASN.1 distinguished name into binary DER-encoded format
  */
@@ -837,6 +870,16 @@
 	    }
 	    else
 	    {
+		if (*src == ',') {
+		   if (*++src == ',') {
+			src--;
+			name.len++;
+			remove_comma(src);
+			break;
+		   } else 
+			src--;
+		}
+
 		name.len -= whitespace;
 		code_asn1_length(name.len, &asn1_name_len);
 
