diff -urNp openswan-2.6.32-cvs-patched/programs/pluto/ike_alg.c openswan-2.6.32-current/programs/pluto/ike_alg.c
--- openswan-2.6.32-cvs-patched/programs/pluto/ike_alg.c	2011-09-01 16:58:25.408675283 -0400
+++ openswan-2.6.32-current/programs/pluto/ike_alg.c	2011-09-29 15:13:08.026292534 -0400
@@ -115,7 +115,7 @@ bool ike_alg_enc_ok(int ealg, unsigned k
 				ealg, key_len);
 		}
 	);
-	if (!ret && *errp)
+	if (!ret && errp)
 		*errp = ugh_buf;
 	return ret;
 }
